### PR TITLE
enh : implement speaker count

### PR DESCRIPTION
**Pull request by uds5501**
Fixes fossasia/open-event-frontend#1936

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The unit tests pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

- [x] All the functions created/modified in this PR contain relevant docstrings.

#### Short description of what this resolves:

Uses speaker_session links which ensures correct speaker counts according to session state within queries itself 

#### Changes proposed in this pull request:

- Add a Session_Speaker_Link Model
- Implement Linking on session and speaker creation
- Use queries to return count instead of set

### CODE REVIEW

**Comment by uds5501** [comment]
@iamareebjamal @CosmicCoder96 Please review :)

**Comment by uds5501** [comment]
A screen shot after the change
image

**Comment by iamareebjamal** [comment]
Instead of this, it should be

```
speakers: {
  accepted:
  confirmed:
  ...
  total
}
```

**Comment by uds5501** [comment]
@iamareebjamal  Understood, shall commit soon

**Comment by uds5501** [comment]
@iamareebjamal does it look fine now?

**Comment by iamareebjamal** [comment]
Build is failing

**Comment by uds5501** [comment]
@iamareebjamal @CosmicCoder96 , my PR doesn't really touch image resizing or anything regarding that, any idea why those particular errors might be showing up?

File: app/api/schema/event_statistics.py [hunk 1]
```diff
@@ -51,7 +61,19 @@ def sessions_rejected_count(self, obj):
         return Session.query.filter_by(event_id=obj.id, state='rejected').count()
 
     def speakers_count(self, obj):
-        return Speaker.query.filter_by(event_id=obj.id).count()
+        accepted = SessionsSpeakersLink.query.filter_by(event_id=obj.id, session_state='accepted').count()
+        confirmed = SessionsSpeakersLink.query.filter_by(event_id=obj.id, session_state='confirmed').count()
+        pending = SessionsSpeakersLink.query.filter_by(event_id=obj.id, session_state='pending').count()
```
Comment (iamareebjamal) [inline app/api/schema/event_statistics.py]: Why not using the functions written?

File: app/api/sessions.py [hunk 2]
```diff
@@ -57,6 +58,15 @@ def after_create_object(self, session, data, view_kwargs):
             send_email_new_session(organizer_email, event_name, link)
             send_notif_new_session_organizer(organizer, event_name, link, session.id)
 
+        if session.speakers != []:
```
Comment (iamareebjamal) [inline app/api/sessions.py]: Not a good way to check for empty list

Comment (iamareebjamal) [reply app/api/sessions.py]: BTW this is a redundant list

File: app/api/sessions.py [hunk 3]
```diff
@@ -57,6 +58,15 @@ def after_create_object(self, session, data, view_kwargs):
...
+            for speaker in session.speakers:
+                ss_link = SessionsSpeakersLink(session_state=session.state,
+                                               session_id=session.id,
+                                               event_id=session.event.id,
+                                               speaker_id=speaker.id)
+                print(ss_link)
```
Comment (iamareebjamal) [inline app/api/sessions.py]: Remove

File: app/api/speakers.py [hunk 4]
```diff
@@ -50,7 +51,6 @@ def before_post(self, args, kwargs, data=None):
                 if not has_access('is_session_self_submitted', session_id=session_id):
                     raise ObjectNotFound({'parameter': 'session_id'},
                                          "Session: {} not found".format(session_id))
-
```
Comment (iamareebjamal) [inline app/api/speakers.py]: Revert unnecessary changes

File: app/api/schema/event_statistics.py (hunk 1 above)
Comment (uds5501) [reply app/api/schema/event_statistics.py]: @iamareebjamal my bad, I thought I removed the functions, I will do that now. They are just redundant

File: app/api/speakers.py (hunk 4 above)
Comment (uds5501) [reply app/api/speakers.py]: Understood, shall do them.

File: app/api/sessions.py (hunk 2 above)
Comment (uds5501) [reply app/api/sessions.py]: @iamareebjamal A `len(session.speakers) != 0` would be fine?

File: app/api/schema/event_statistics.py (hunk 1 above)
Comment (iamareebjamal) [reply app/api/schema/event_statistics.py]: Created a single function which takes a parameter which varies as accepted, pending, etc

File: app/api/sessions.py (hunk 2 above)
Comment (iamareebjamal) [reply app/api/sessions.py]: No need. Directly use the loop

Comment (uds5501) [reply app/api/sessions.py]: understood!

File: app/api/speakers.py (hunk 4 above)
Comment (uds5501) [reply app/api/speakers.py]: Done ! :)

File: app/api/sessions.py [hunk 5]
```diff
@@ -57,6 +58,13 @@ def after_create_object(self, session, data, view_kwargs):
             send_email_new_session(organizer_email, event_name, link)
             send_notif_new_session_organizer(organizer, event_name, link, session.id)
 
+        for speaker in session.speakers:
+            ss_link = SessionsSpeakersLink(session_state=session.state,
```
Comment (CosmicCoder96) [inline app/api/sessions.py]: ss_link is not a good variable name.

File: app/api/sessions.py [hunk 6]
```diff
@@ -149,6 +157,25 @@ def after_update_object(self, session, data, view_kwargs):
...
                                                  session.state, link, session.id)
+        if 'state' in data:
+            entry_count = SessionsSpeakersLink.query.filter_by(session_id=session.id)
+            if entry_count.count() == 0:
+                is_patch_request = False
+            else:
+                is_patch_request = True
+
+            if is_patch_request:
+                for sess in entry_count:
```
Comment (CosmicCoder96) [inline app/api/sessions.py]: `sess` is not clear, improve it.

**Comment by uds5501** [comment]
@CosmicCoder96 reviews acknowledged

**Comment by uds5501** [comment]
@CosmicCoder96 @iamareebjamal this PR is now ready for review, although I may need some assistance regarding unpacking this speaker count into models in the frontend

**Comment by iamareebjamal** [comment]
Where's the migration file in the PR?

**Comment by uds5501** [comment]
@iamareebjamal added the migration file.

File: migrations/versions/35f427e85075_.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+import sqlalchemy as sa
+import sqlalchemy_utils
+from sqlalchemy.dialects import postgresql
+
+# revision identifiers, used by Alembic.
+revision = '35f427e85075'
+down_revision = '565ea5bc3937'
+
+
+def upgrade():
```
Comment (iamareebjamal) [inline migrations/versions/35f427e85075_.py]: It's a reverse migration

File: app/api/schema/event_statistics.py [hunk 8]
```diff
@@ -50,8 +51,23 @@ def sessions_pending_count(self, obj):
     def sessions_rejected_count(self, obj):
         return Session.query.filter_by(event_id=obj.id, state='rejected').count()
 
+    def speakers_count_type(self, obj, state='pending'):
```
Comment (CosmicCoder96) [inline app/api/schema/event_statistics.py]: @uds5501 why is the default state pending?
wouldn't it be better to not have default as pending, rather it's better if an error is thrown if state is not supplied as the function is basically returning sessions by states? Pending seems arbitrary.

Comment (uds5501) [reply app/api/schema/event_statistics.py]: @CosmicCoder96 Actually I think this function isn't being used anywhere at all so I will be removing it in the next commit
